intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+use names to automatically check that APIs are being used correctly at runtime, providing extra
+safety. Names can also be used to rearrange dimensions, for example, to support
+"broadcasting by name" rather than "broadcasting by position".
+
+this tutorial is intended as a guide to the functionality that will
+be included with the 1.3 launch. By the end of it, you will be able to:
+
+- Initiate a ``Tensor`` with named dimensions, as well as removing or renmaing those dimensions
+- Understand the basics of how dimension names are propagated through operations
+- See how naming dimensions enables clearer code in two key areas:
+    - Broadcasting operations
+    - Flattening and unflattening dimensions
+
+Finally, we'll put this into practice by writing a multi-headed attention module
+using named tensors.
+
+Named tensors in PyTorch are inspired by and done in collaboration with
+`Sasha Rush <https://tech.cornell.edu/people/alexander-rush/>`_.
+The original idea and proof of concept were proposed in his
+`January 2019 blog post <http://nlp.seas.harvard.edu/NamedTensor>`_.
+"""
+
+######################################################################
+# Basics: named dimensions
+# ------------------------
+#
+# PyTorch now allows Tensors to have named dimensions; factory functions
+# now take a new `names` argument that associates a name with each dimension.
+# This works with most factory functions, such as
+#
+# - `tensor`
+# - `empty`
+# - `ones`
+# - `zeros`
+# - `randn`
+# - `rand`
+#
+# Here we construct a tensor with names:
+
+import torch
+imgs = torch.randn(1, 2, 2, 3 , names=('N', 'C', 'H', 'W'))
+
+# Unlike in
+# `the original named tensors blogpost <http://nlp.seas.harvard.edu/NamedTensor>`_,
+# named dimensions are ordered: `tensor.names[i]` is the name of the `i`th dimension of `tensor`.
+#
+# There are two ways rename a ``Tensor``'s dimensions:
+#
+
+print(imgs.names)
+
+# Method #1: set .names attribute
+imgs.names = ['batch', 'channel', 'width', 'height']
+print(imgs.names)
+
+# Method #2: specify new names:
+imgs.rename(channel='C', width='W', height='H')
+print(imgs.names)
+
+######################################################################
+# The preferred way to remove names is to call ``tensor.rename(None)``:
+
+imgs.rename(None)
+print(imgs.names)
+
+######################################################################
+# Unnamed tensors (tensors with no named dimensions) still work as normal and do
+# not have names in their `repr`.
+
+unnamed = torch.randn(2, 1, 3)
+print(unnamed)
+print(unnamed.names)
+
+######################################################################
+# Named tensors do not require that all dimensions be named.
+
+imgs = torch.randn(3, 1, 1, 2, names=('N', None, None, None))
+print(imgs.names)
+
+######################################################################
+# Because named tensors coexist with unnamed tensors, we need a nice way to write named-tensor-aware

```

Comment:
nit: It might be good to stay within 80 characters per line so that someone reading this tutorial directly in the code won't see weird line breaks.